src/test/java/com/google/devtools/build/lib/remote/blobstore/http/HttpBlobStoreTest.java

```diff
@@ -54,73 +63,314 @@
+                          if (UnixDomainServer.this.handler != null) {
+                            UnixDomainServer.this.handler.channelReadComplete(ctx);
+                          }
+                        }
+
+                        @Override
+                        public void userEventTriggered(ChannelHandlerContext ctx,
+                            Object o) throws Exception {
+                          if (UnixDomainServer.this.handler != null) {
+                            UnixDomainServer.this.handler.userEventTriggered(ctx, o);
+                          }
+                        }
+
+                        @Override
+                        public void channelWritabilityChanged(
+                            ChannelHandlerContext ctx) throws Exception {
+                          if (UnixDomainServer.this.handler != null) {
+                            UnixDomainServer.this.handler.channelWritabilityChanged(ctx);
+                          }
+                        }
+
+                        @Override
+                        public void handlerAdded(ChannelHandlerContext ctx)
+                            throws Exception {
+                          if (UnixDomainServer.this.handler != null) {
+                            UnixDomainServer.this.handler.handlerAdded(ctx);
+                          }
+                        }
+
+                        @Override
+                        public void handlerRemoved(ChannelHandlerContext ctx)
+                            throws Exception {
+                          if (UnixDomainServer.this.handler != null) {
+                            UnixDomainServer.this.handler.handlerRemoved(ctx);
+                          }
+                        }
+
+                        @Override
+                        public void exceptionCaught(ChannelHandlerContext ctx,
+                            Throwable throwable) throws Exception {
+                          if (UnixDomainServer.this.handler != null) {
+                            UnixDomainServer.this.handler.exceptionCaught(ctx, throwable);
+                          }
+                        }
+                      });
+                    }
+                  });
+      try {
+        ServerChannel actual = ((ServerChannel) sb.bind(newDomainSocketAddress()).sync().channel());
+        this.serverChannel = mock(ServerChannel.class, AdditionalAnswers.delegatesTo(actual));
+      } catch (Exception e) {
+        throw new IllegalStateException(e);
+      }
+
+    }
+
+    public ServerChannel start(ChannelInboundHandler handler) {
+      reset(this.serverChannel);
+      this.handler = handler;
+      return this.serverChannel;
+    }
+
+    public void stop(ServerChannel serverChannel) {
+      // return an invalid path otherwise attempted connections will likely get
+      // read timeouts intead of connection timeouts
+      when(this.serverChannel.localAddress()).thenReturn(new DomainSocketAddress(""));
```

Comment:
Can you elaborate what this is doing please? I don't think I understand :-).

Reply:
I've improved the comment. It's a big of an odd hack, no doubt! Let me know if you want more info.